components/zip321/src/lib.rs

```diff
@@ -186,29 +199,22 @@ impl Payment {
     pub fn memo(&self) -> Option<&MemoBytes> {
         self.memo.as_ref()
     }
+}
 
-    /// A human-readable label for this payment within the larger structure
-    /// of the transaction request.
-    pub fn label(&self) -> Option<&String> {
-        self.label.as_ref()
-    }
-
-    /// A human-readable message to be displayed to the user describing the
-    /// purpose of this payment.
-    pub fn message(&self) -> Option<&String> {
-        self.message.as_ref()
-    }
-
-    /// A list of other arbitrary key/value pairs associated with this payment.
-    pub fn other_params(&self) -> &[(String, String)] {
-        self.other_params.as_ref()
-    }
+/// Split zatoshis into integer ZEC and fraction remainder zatoshis.
+fn split_zatoshis(amount: Zatoshis) -> (u64, u64) {
+    let coins = u64::from(amount) / zcash_protocol::value::COIN;
+    let zats = u64::from(amount) % zcash_protocol::value::COIN;
+    (coins, zats)
+}
 
-    /// A utility for use in tests to help check round-trip serialization properties.
-    #[cfg(any(test, feature = "test-dependencies"))]
-    pub(crate) fn normalize(&mut self) {
-        self.other_params.sort();
-    }
+/// Combine ZEC and a fractional remainder of zatoshis.
+fn combine_zatoshis((zec, zats): (u64, u64)) -> Result<Zatoshis, Zip321Error> {
+    zec.checked_mul(COIN)
+        .and_then(|coin_zats| coin_zats.checked_add(zats))
+        .ok_or(BalanceError::Overflow)
+        .and_then(Zatoshis::from_u64)
+        .map_err(|_| Zip321Error::Other(format!("Not a valid zat amount: {zec}.{zats}")))
```

Comment:
I lifted this logic out of the parser, which now keeps ZEC and the fractional zats separately. `zip321::Payment` still contains `Zatoshis` and uses this logic for validation.